### PR TITLE
[docs] [WIP] separate gulp tasks by file

### PR DESCRIPTION
**Pull request by darlanmendonca**
Hey guys, following the discussion in https://testcafe-discuss.devexpress.com/t/gulp-tasks-by-file/51, I make this PR, to suggest a better organization to gul tasks
- one task by file

this help to create little scopes, instead a large gulpfile with zillions of files

another thing, each task use a named function instead a anonymous. this help to debug errors.
I named each function with a suffix Task, so task clean, execute the function cleanTask. The suffix is necessary to when the task has a reserved name, e.g. a task called default, will conflict with reserved word default in javascript, so, the function defaultTask avoid that.

I add tasks/*.js to task lint, and fix all errors, but, for really works all, we need test if I dont forget rename a path to file, example `require('test/path/to/file`, now need begin to `../`, because path is relative to the current file.

variables used across tasks, I put all inside a tasks/config.js, in the future maybe rename the variable can be cool.

For now, I will named this PR for WIP (work in progress), because we need test each task, to avoid possibles errors :D

### CODE REVIEW

**Comment by darlanmendonca** [comment]
anyone can help me with the CI? I dont know if is an error in my files, or if is anything with .babelrc (CI instable?)

**Comment by inikulin** [comment]
@darlanmendonca Indeed, the issue is very odd. But looking at `/home/travis/build/testcafe-build-bot/testcafe/tasks/src/client/.babelrc` path in error message makes me think that the issue is in the incorrect source and destination paths of gulp task: previously all paths were relative to the Gulpfile, but now seems like gulp tries to resolve them relative to the task file path.

**Comment by inikulin** [comment]
@darlanmendonca Does this problem reproduces if you run tasks locally?

**Comment by darlanmendonca** [comment]
@inikulin thanks, you're right, in client-scripts-bundle.js:34 I have a relative path that I dont update.
I fix, and now find for relative paths that I forget

**Comment by inikulin** [comment]
I don't have access to my computer at the moment, so If you wouldn't mind I'll review the rest tomorrow.

**Comment by darlanmendonca** [comment]
yep, when you can of course :D

**Comment by darlanmendonca** [comment]
from "Does this problem reproduces if you run tasks locally" @inikulin,

I try run the tasks that I know locally, but has many tasks in different environtments, etc.
I need a help of team, or who know the tasks, to a nice code review, and test tasks

**Comment by AlexanderMoskovkin** [comment]
@darlanmendonca FYI: The `test-server` task [is failed](https://travis-ci.org/testcafe-build-bot/testcafe/jobs/170513198#L1620) but it doesn't relate to your commits. It became failed since some `babel` deep dependencies were updated. I'll fix this soon.

**Comment by darlanmendonca** [comment]
ok, thanks @AlexanderMoskovkin

**Comment by AlexanderMoskovkin** [comment]
> I'll fix this soon.

done in github.com/DevExpress/testcafe/pull/914
You should rebase your PR to upstream/master

File: tasks/build-website-development.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+var gulp = require('gulp');
+var config = require('./config.js');
+
+gulp.task('build-website-development', ['prepare-website'], buildWebsiteEevelopmentTask);
+
+function buildWebsiteEevelopmentTask (cb) {
```
Comment (inikulin) [inline tasks/build-website-development.js]: Here and in other places: we don't need intermediate named function here, just pass it directly to `gulp.task`

File: tasks/build-website-testing.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+var gulp = require('gulp');
+var config = require('./config.js');
```
Comment (inikulin) [inline tasks/build-website-testing.js]: Our code style requires assignments to be equally aligned:
```js
var gulp    = require('gulp');
var config = require('./config.js');
```
Unfortunately I'm not aware about eslint rule that enforces this behavior. Maybe you could help?

File: tasks/config.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        // NOTE: https://github.com/DevExpress/testcafe/issues/471
+        // problem with extra scroll reproduced only on saucelabs
+        // virtual machines with ios device emulators
+        version:     '9.3',
+        deviceName:  'iPhone 6 Plus'
+    }
+];
+
+
+function buildWebsite (mode, cb) {
```
Comment (inikulin) [inline tasks/config.js]: I have a feeling that this helper functions has nothing to do with config, maybe move it to `helpers.js` file?

File: Gulpfile.js [hunk 4]
```diff
@@ -1,35 +1,6 @@
...
-var Promise              = require('pinkie');
-var markdownlint         = require('markdownlint');
-var prompt               = require('gulp-prompt');
-var nodeVer              = require('node-version');
-var functionalTestConfig = require('./test/functional/config');
-var assignIn             = require('lodash').assignIn;
-var runSequence          = require('run-sequence');
-var yaml                 = require('js-yaml');
-
+var ll = require('gulp-ll');
```
Comment (inikulin) [inline Gulpfile.js]: It would be nice to separate task files by directories, e.g.: `lint`, `website`, `functional-tests`, `server-tests`, etc.

**Comment by darlanmendonca** [comment]
@AlexanderMoskovkin how I can rebase? with

```sh
git rebase upstream/gulp-tasks-by-file
```

?

**Comment by inikulin** [comment]
@darlanmendonca 

> @AlexanderMoskovkin how I can rebase? with


In your local branch directory:
```sh
git remote add upstream https://github.com/DevExpress/testcafe.git
```
Now you have upstream remote repo setup that points to this one. The next step is to fetch upstream changes:
```sh
git fetch upstream
```
and then perform rebase (there may be merge conflicts during rebase, you can resolve them as described [here](https://help.github.com/articles/resolving-merge-conflicts-after-a-git-rebase/):
```sh
git rebase upstream/master
```
Then update PR with rebased version:
```sh
git push -f
```

**Comment by darlanmendonca** [comment]
thanks @inikulin, rebased.

File: tasks/build-website-development.js (hunk 1 above)
Comment (darlanmendonca) [inline tasks/build-website-development.js]: I dont understand, you talk about middlewares (array of tasks), or named the callback task (the function in last argument)?

File: tasks/build-website-testing.js (hunk 2 above)
Comment (darlanmendonca) [inline tasks/build-website-testing.js]: probaly is a bug in eslint with that rule (key-spacing align value)

to fix that, only opening a issue in eslint, I believe.

as this standard, I dont like this, because cause hard diffs in code-review/github, but it is just my opnion

for now, I will change all variables when eslint dont can validate, to follow the pattern

Comment (inikulin) [inline tasks/build-website-testing.js]: > as this standard, I dont like this, because cause hard diffs in code-review/github, but it is just my opnion

Yeah, it's a common opinion. But from our experience it's not a big deal when it comes to diffs. But on the other hand, variable definitions are more readable.

Comment (darlanmendonca) [inline tasks/build-website-testing.js]: yeah, it has pros and cons.

for now, I change to follow the rule right?

Comment (darlanmendonca) [inline tasks/build-website-testing.js]: even with the rule somewhat bug in eslint

File: tasks/build-website-development.js (hunk 1 above)
Comment (inikulin) [inline tasks/build-website-development.js]: I mean we can do:

```js
gulp.task('build-website-development', ['prepare-website'], function (cb) {
     config.buildWebsite('development', cb);
});
```

Comment (darlanmendonca) [inline tasks/build-website-development.js]: I suggest named functions, lets suppose you dont have a variable config, or another error inside that function, if is a anonymous function, the error will be simillar to:

```
error on anonimous function in file.js
```

now if we name the function

```
error on 'methodName' in file.js
```

File: tasks/build-website-testing.js (hunk 2 above)
Comment (inikulin) [inline tasks/build-website-testing.js]: > for now, I change to follow the rule right?

Yes

Comment (darlanmendonca) [inline tasks/build-website-testing.js]: ok, the right is:

```js
var gulp =   require('gulp');
var config = require('./config.js');
``
or 

```js
var gulp    = require('gulp');
var config = require('./config.js');
```

reading the key-spacing align value, I believe the rule get only values inside a object, not variable assigns, because of that rule is called key and not variable-spacing :D

Eslint probaly not can lint our code because we dont have a rule in .eslint, at now, to force this, I believe

File: Gulpfile.js [hunk 5]
```diff
@@ -42,527 +13,6 @@ ll
...
-
-gulp.task('test-functional-local', ['build'], function () {
-    return testFunctional('test/functional/fixtures', functionalTestConfig.testingEnvironmentNames.localBrowsers);
-});
-
-gulp.task('test-functional-travis-legacy', ['build'], function () {
-    return testFunctional('test/functional/legacy-fixtures', functionalTestConfig.testingEnvironmentNames.legacy);
-});
+// require all tasks inside ./task directory
+requireAll(path.join(__dirname, '/tasks/'));
```
Comment (inikulin) [inline Gulpfile.js]: @darlanmendonca There is an issue with requiring eslint on node 10, it uses ES6 syntax. I guess we can just do:
```js
var nodeVersion = require('node-version');

if(Number(nodeVersion) < 4) 
    require('../tasks/test-server');
else
    requireAll(path.join(__dirname, '/tasks/'));
```

File: tasks/build-website-testing.js (hunk 2 above)
Comment (inikulin) [inline tasks/build-website-testing.js]: > if I commit it now, nothing ensures that I dont forget anyone, or give a wrong space before =

Yes, this makes sense. Let's skip this issue for now.

File: Gulpfile.js (hunk 5 above)
Comment (darlanmendonca) [inline Gulpfile.js]: nodeVersion return a object, so, we wish

```js
if(Number(nodeVersion.major) < 4) 
    require('../tasks/test-server');
else
    requireAll(path.join(__dirname, '/tasks/'));
```

I just dont understand the directory test-server, this be related to create subdirectories inside tasks? (comment more above)?

Comment (inikulin) [inline Gulpfile.js]: Oh, my bad, your code snippet is correct.

> I just dont understand the directory test-server, this be related to create subdirectories inside tasks? (comment more above)?

Yes.

**Comment by inikulin** [comment]
@darlanmendonca Hi, any updates on this one? We are really close to finish it.

**Comment by darlanmendonca** [comment]
:D oh yeah, let's go

File: Gulpfile.js (hunk 5 above)
Comment (darlanmendonca) [inline Gulpfile.js]: @inikulin, I'm thinking about it too

I believe that is better to do it in a new PR, so we can create a new scope to talk about a subject in specific, like "subdivide the test-server tasks  or "turn redundant tasks in just one" (test-client-legacy-travis-mobile, test-client-legacy, test-client-old-browsers-travis are examples).

My point with this PR is just create little scopes by file/task, to allow us see better each task, instead has all in a single file (gulpfile)

Comment (darlanmendonca) [inline Gulpfile.js]: if the all tasks run without bugs, we can finish this PR, and begin to see in details a task, or group of tasks, to improve it

Comment (inikulin) [inline Gulpfile.js]: OK, let's do this in separate PR.

**Comment by darlanmendonca** [comment]
any idea why task 'travis' broken in CI?

I see  in log an error using gulp-eslint module writed and exported in es6
I will check the env GULP_TASK

**Comment by inikulin** [comment]
It's because `build` task has `lint` as a dependency. My bad, I've mislead you with https://github.com/DevExpress/testcafe/pull/919#discussion_r85299549 

We need to modify `lint.js` the way that it will require `eslint` and actually do anything only in node `>=4.0.0`.

**Comment by darlanmendonca** [comment]
humm, I think is better do a PR to gulp-eslint to add support in es5
for now we can do this, only require gulp-eslint in node > 4

but if gulp-eslint accept our PR, we back and remove this "conditional requirement" from our lint.js

**Comment by inikulin** [comment]
I believe they will not accept a PR, since eslint itself dropped support for node 0.10

**Comment by darlanmendonca** [comment]
humm, ok
if they dont want add support to node 0.10 (with babel transpile), we can continue with conditional require

I push, waiting CI to tests

**Comment by darlanmendonca** [comment]
@inikulin know the problem in CI? The number of free machines (2) is less than requested (3)

**Comment by AlexanderMoskovkin** [comment]
@darlanmendonca Yes, it were no enough available machines to run the tests on saucelabs. I'll restart them

**Comment by AlexanderMoskovkin** [comment]
Guys, I'll add the [docs] prefix to the PR and rerun tests to be sure that documentation tests are ok

**Comment by AlexanderMoskovkin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
It's something wrong with documentation tests. Maybe @VasilyStrelyaev can help us to determine the cause of the problem?
https://travis-ci.org/testcafe-build-bot/testcafe/builds/172243777

File: tasks/config.js (hunk 3 above)
Comment (AlexanderMoskovkin) [inline tasks/config.js]: I agree with @inikulin here. Let's refactor this

Comment (darlanmendonca) [inline tasks/config.js]: what I do?
with websiteServer in line 142?

Comment (darlanmendonca) [inline tasks/config.js]: @AlexanderMoskovkin what I do with websiteServer in line 142?

Comment (AlexanderMoskovkin) [inline tasks/config.js]: I propose to move functions for website serving to the `website.js` file:
```js
// website.js

/* some imports here */

var websiteServer = null;

function serve (cb) {
    var app = connect().use('/testcafe', serveStatic('site/deploy'));

    config.websiteServer = app.listen(8080, cb);
}

function build () {
    // ...
}

function test () {
    // ...
}

module.exports = {
    serve: serve,
    build: build,
    test: test
};
```
In this case the `serve-website.js` will be
```js
var website = require('./website.js');

gulp.task('serve-website', website.serve);
```

Comment (AlexanderMoskovkin) [inline tasks/config.js]: Also I propose to move the `testFunctional` function to a separate `test-functional.js` file. In this case `config.js` contains only settings for client tests.

And thanks for your great job!

**Comment by AlexanderMoskovkin** [comment]
Hi @darlanmendonca, how is your progress? Do you need some assistants from us?

**Comment by darlanmendonca** [comment]
@AlexanderMoskovkin sorry by delay, running with a work these days. To night I will have time, then I continue to we can complete this PR :D

**Comment by darlanmendonca** [comment]
@AlexanderMoskovkin looking now, anyone can help me? 
the task test-website-travis broke because cant access close method of websiteServer (config.js:142)

**Comment by AlexanderMoskovkin** [comment]
@darlanmendonca Yes, I've proposed how we can refactor this in my [comment](https://github.com/DevExpress/testcafe/pull/919/files#r85959227)

**Comment by kirovboris** [comment]
I will close the pull request, because no activity has been encountered for a long time.
